tools/xdp/s2n-quic-xdp/src/ring/cursor.rs

```diff
@@ -0,0 +1,488 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    if_xdp::{RingFlags, RingOffsetV2},
+    mmap::Mmap,
+};
+use core::{
+    ffi::c_void,
+    fmt,
+    marker::PhantomData,
+    num::Wrapping,
+    ptr::NonNull,
+    sync::atomic::{AtomicU32, Ordering},
+};
+
+/// A structure for tracking a ring shared between a producer and consumer
+///
+/// See [xsk.h](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/headers/xdp/xsk.h#L34-L42).
+#[derive(Debug)]
+pub struct Cursor<T: Copy + fmt::Debug> {
```

Comment:
I want to preserve the `desc` rather than `ring`. Ring is just used all over the place and I wanted to say that this points to the descriptors.

I can add the docs and reorder, though.